core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java

```diff
@@ -94,54 +88,64 @@ private String toPolicyLocation(ComponentLocation componentLocation) {
     return from(publisher)
         .doOnNext(coreEvent -> logExecuteNextEvent("Before execute-next", coreEvent.getContext(),
                                                    coreEvent.getMessage(), muleContext.getConfiguration().getId()))
-        .flatMap(event -> {
-          PolicyStateId policyStateId = stateIdFactory.create(event);
-          Processor nextOperation = policyStateHandler.retrieveNextOperation(policyStateId.getExecutionIdentifier());
-
-          if (nextOperation == null) {
-            return error(new MuleRuntimeException(createStaticMessage("There's no next operation configured for event context id "
-                + policyStateId.getExecutionIdentifier())));
-          }
-
+        .map(event -> (CoreEvent) policyEventConverter.createEvent(saveState((PrivilegedEvent) event),
+                                                                   getOriginalEvent(event)))
+        .doOnNext(event -> {
           popBeforeNextFlowFlowStackElement().accept(event);
           notificationHelper.notification(BEFORE_NEXT).accept(event);
+        })
+
+        .flatMap(event -> subscriberContext().flatMap(ctx -> Mono.just(event).transform(ctx.get(POLICY_NEXT_OPERATION)))
+            .cast(CoreEvent.class)
+            .doOnError(MessagingException.class, this::handleExceptionInNext))
+
+        // TODO MULE-16371 Replace the flatMap above with this compose. Review how to handle the error from the flow in source
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1135))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7629&issues=AWiGlnYDotIXnmsUonib&open=AWiGlnYDotIXnmsUonib)

[ISSUE_KEY]: AWiGlnYDotIXnmsUonib